cylc/flow/task_pool.py

```diff
@@ -1477,304 +1450,530 @@ def can_spawn(self, name: str, point: 'PointBase') -> bool:
             LOG.debug(
                 'Not spawning %s/%s: before initial cycle point', point, name)
             return False
-        elif self.config.final_point and point > self.config.final_point:
+
+        if self.config.final_point and point > self.config.final_point:
             # Only happens on manual trigger beyond FCP
             LOG.debug(
                 'Not spawning %s/%s: beyond final cycle point', point, name)
             return False
+
+        # Is it on-sequence and within recurrence bounds.
+        if not self.config.get_taskdef(name).is_valid_point(point):
+            LOG.warning(
+                self.ERR_PREFIX_TASK_NOT_ON_SEQUENCE.format(
+                    name, point
+                )
+            )
+            return False
+
         return True
 
+    def _get_task_history(self, name, point, flow_nums):
+        """Get details of previous submits for this task.
+
+        """
+        info = self.workflow_db_mgr.pri_dao.select_prev_instances(
+            name, str(point)
+        )
+        try:
+            submit_num = max(s[0] for s in info)
+        except ValueError:
+            # never spawned before in any flow
+            submit_num = 0
+
+        prev_completed = False  # did not complete in the flow
+        prev_flow_wait = False  # did not wait in the flow
+        # orig_fnums = set()
+
+        for snum, f_wait, old_fnums, is_complete in info:
+            if set.intersection(flow_nums, old_fnums):
+                # matching flows
+                prev_completed = is_complete
+                prev_flow_wait = f_wait
+                # orig_fnums = old_fnums
+                if not prev_completed:
+                    # There may be multiple entries with flow overlap due
+                    # to merges (they'll have have same snum and f_wait);
+                    # keep going to find the complete one, if any .
+                    continue
+                LOG.warning(
+                    f"{point}/{name} already completed for flow"
+                    f" {stringify_flow_nums(flow_nums, full=True)} via"
+                    f" {point}/{name}/{snum:02d}"
+                    f"{stringify_flow_nums(old_fnums, full=True)}"
+                )
+                break
+
+        return submit_num, prev_completed, prev_flow_wait
```

Comment:
For 8.2 workflows that have been restarted with 8.3, the default value for `is_complete` is `0`. This means that complete tasks may be loaded back as incomplete.

It might be reasonable to say this is a known limitation, maybe a couple tasks get re-run after restart, so what. If so we need to document this clearly.

However, it might be reasonable to build in some back-compat here, e.g. get the upgrader to put in a value of `2` or something like that to mean "unknown", and re-evaluate completion using the new config.

Reply:
I can test it with this (run with 8.2.4, restart with this branch - task `c` should not re-run):
```ini
[scheduling]
    [[graph]]
           R1 = "a => b => c => d"
[runtime]
    [[a]]
        script = """
           cylc trigger --wait $CYLC_WORKFLOW_ID//1/c
           sleep 3
           cylc stop $CYLC_WORKFLOW_ID
        """
    [[b,c,d]]
```
It currently does the right thing in this case, but for the wrong reason - on fixing that, it does rerun `c` as you suggest.

Reply:
Fixed by defaulting to 2 for unknown completion (to cause a re-evaluation).

Reply:
Probably not worth adding a test for this?

It's a very niche problem that users are unlikely to hit now, and will never hit once on 8.3.

 (8.2 to 8.3 upgrade-restart that encounters a flow-wait task)